tests/Catch2_Tests/comparators/ComparatorDisposerTest.cpp

```diff
@@ -0,0 +1,56 @@
+#include "catch2/catch.hpp"
+#include "ApprovalTests/comparators/ComparatorDisposer.h"
+#include "Approvals.h"
+#include <fstream>
+#include <iterator>
+#include <memory>
+
+using namespace ApprovalTests;
+
+namespace
+{
+    class TestComparator : public ApprovalComparator
+    {
+    public:
+        bool contentsAreEquivalent(std::string receivedPath,
+                                   std::string approvedPath) const override
+        {
+            std::ifstream receivedStream(receivedPath.c_str());
+            std::ifstream approvedStream(approvedPath.c_str());
+
+            using InputIter = std::istreambuf_iterator<char>;
+
+            auto receivedText =
+                std::string(InputIter{receivedStream}, InputIter{});
+            auto approvedText =
+                std::string(InputIter{approvedStream}, InputIter{});
+
+            return (receivedText == "foo\n") && (approvedText == "bar\n");
+        }
+    };
+}
+
+TEST_CASE("Type traits of ComparatorDisposer")
+{
+
+    REQUIRE_FALSE(std::is_copy_constructible<ComparatorDisposer>::value);
+    REQUIRE_FALSE(std::is_copy_assignable<ComparatorDisposer>::value);
+
+    REQUIRE(std::is_move_constructible<ComparatorDisposer>::value);
+    REQUIRE_FALSE(std::is_move_assignable<ComparatorDisposer>::value);
+}
+
+TEST_CASE("ComparatorDisposer object can be moved without disposing a "
+          "comparator twice")
+{
+    std::unique_ptr<ComparatorDisposer> disposer;
+
+    {
+        auto temporaryDisposer = FileApprover::registerComparatorForExtension(
+            ".txt", std::make_shared<TestComparator>());
+
+        disposer.reset(new ComparatorDisposer(std::move(temporaryDisposer)));
+    }
+
+    REQUIRE_NOTHROW(Approvals::verify("foo"));
```

Comment:
If this fails, users see a diffing tool, and this output:

```
/Users/clare/Documents/develop/ApprovalTests/ApprovalTests.cpp/tests/Catch2_Tests/comparators/ComparatorDisposerTest.cpp:51: FAILED:
  REQUIRE_NOTHROW( Approvals::verify("foo") )
due to unexpected exception with message:
  Failed Approval: 
  Received does not match approved 
  Received : "/Users/clare/Documents/develop/ApprovalTests/ApprovalTests.cpp/
  tests/Catch2_Tests/comparators/approval_tests/ComparatorDisposerTest.
  ComparatorDisposer_object_can_be_moved_without_disposing_a_comparator_twice.
  received.txt" 
  Approved : "/Users/clare/Documents/develop/ApprovalTests/ApprovalTests.cpp/
  tests/Catch2_Tests/comparators/approval_tests/ComparatorDisposerTest.
  ComparatorDisposer_object_can_be_moved_without_disposing_a_comparator_twice.
  approved.txt"
```

In that situation, someone who didn't understand this test might be inclined to approve the new output, at which point the test would always pass, whilst not actually testing the intended behaviour...

One way to detect and prevent this would be to repeat the `REQUIRE_NOTHROW(Approvals::verify("foo"));`, both before and after the `disposer.reset` line...

Reply:
I changed the test to use `FileApprover::verify` instead of `Approvals::verify` so that the approved file does not need to be put in git and diff tool is not run.